play-services-base/src/main/java/org/microg/gms/common/MultiConnectionKeeper.java

```diff
@@ -21,32 +21,142 @@
+    }
+
+    private Boolean isGoogleOrMicrogSig(PackageManager pm, String packageId) throws PackageManager.NameNotFoundException {
+        List<String> signatures = new LinkedList<>(Arrays.asList(GOOGLE_PRIMARY_KEYS));
+        signatures.addAll(Arrays.asList(MICROG_PRIMARY_KEYS));
+        return signatureIsIn(pm, packageId, signatures);
+    }
+
+    private Boolean isMicrogSig(PackageManager pm, String packageId) throws PackageManager.NameNotFoundException {
+        List<String> signatures = Arrays.asList(MICROG_PRIMARY_KEYS);
+        return signatureIsIn(pm, packageId, signatures);
+    }
+
+    private Boolean signatureIsIn(PackageManager pm, String packageId, List<String> signatures) throws PackageManager.NameNotFoundException {
+        Signature[] appSignatures = pm.getPackageInfo(packageId, PackageManager.GET_SIGNATURES).signatures;
+        for (Signature sig : appSignatures) {
+            if (sig != null && signatures.contains(sha1sum(sig.toByteArray())))
+                return true;
+        }
+        return false;
+    }
+
+    private String sha1sum(byte[] bytes) {
+        MessageDigest md;
+        try {
+            md = MessageDigest.getInstance("SHA1");
+        } catch (final NoSuchAlgorithmException e) {
+            return null;
+        }
+        if (md != null) {
+            bytes = md.digest(bytes);
+            if (bytes != null) {
+                StringBuilder sb = new StringBuilder(2 * bytes.length);
+                for (byte b : bytes) {
+                    sb.append(String.format("%02x", b));
+                }
+                return sb.toString();
+            }
+        }
+        return null;
+    }
+
+    private String getTargetPackageWithoutPref() {
+        // Pref: gms > microG > self
+        PackageManager pm = context.getPackageManager();
+        try {
+            if (isSystem(pm, GMS_PACKAGE_NAME) || isGoogleOrMicrogSig(pm, GMS_PACKAGE_NAME)) {
+                Log.d(TAG, GMS_PACKAGE_NAME + " found !");
+                return GMS_PACKAGE_NAME;
+            } else {
+                Log.w(TAG, GMS_PACKAGE_NAME + " found with another signature");
+            }
+        } catch (PackageManager.NameNotFoundException e) {
+            Log.d(TAG, GMS_PACKAGE_NAME + " not found");
+        }
+        try {
+            if (isMicrogSig(pm, MICROG_PACKAGE_NAME)) {
+                Log.d(TAG, MICROG_PACKAGE_NAME + " found !");
+                return MICROG_PACKAGE_NAME;
+            } else {
+                Log.w(TAG, MICROG_PACKAGE_NAME + " found with another signature");
+            }
+        } catch (PackageManager.NameNotFoundException e) {
+            Log.d(TAG, MICROG_PACKAGE_NAME + " not found");
+        }
+        return null;
+    }
+
+    private String getTargetPackage() {
+        SharedPreferences prefs = context.getSharedPreferences(PREF_MASTER, Context.MODE_PRIVATE);
+        final String SELF = "SELF";
```

Comment:
Instead of special handling for `SELF` / `null`, you can just use `context.getPackageName()` (both in the preference and returned from `getTargetPackage()` and extend `isMicrog()` to return true if the service's package name is the current package name (because the service won't be found inside the same package if it's not microG)